include/chainbase/shared_cow_vector.hpp

```diff
@@ -0,0 +1,266 @@
+
+      template<class I, std::enable_if_t<std::is_constructible_v<T, I>, int> = 0 >
+      shared_cow_vector& operator=(const std::vector<I>& v) {
+         clear_and_construct(v.size(), 0, [&](T* dest, std::size_t idx) {
+            std::construct_at(dest, v[idx]);
+         });
+         return *this;
+      }
+
+      template<class I, std::enable_if_t<std::is_constructible_v<T, I>, int> = 0 >
+      shared_cow_vector& operator=(std::initializer_list<I> init) {
+         clear_and_construct(init.size(), 0, [&](T* dest, std::size_t idx) {
+            std::construct_at(dest, std::data(init)[idx]);
+         });
+         return *this;
+      }
+
+      void clear() {
+         dec_refcount();
+         _data = nullptr;
+      }
+
+      template<typename F>
+      void clear_and_construct(std::size_t new_size, std::size_t copy_size, F&& f) {
+         assert(copy_size <= new_size);
+         assert(copy_size == 0 || (_data && copy_size <= _data->size));
+         
+         if (_data && _data->reference_count == 1 && _data->size == new_size)
+            std::destroy(_data->data + copy_size, _data->data + new_size);
+         else {
+            _alloc<false>(data(), new_size, copy_size); // construct == false => uninitialized memory
+         }
+         for (std::size_t i=copy_size; i<new_size; ++i)
+            static_cast<F&&>(f)(_data->data + i, i); // `f` should construct objects in place 
+      }
+
+      void assign(const T* ptr, std::size_t size) {
+         if (_data && _data->reference_count == 1 && _data->size == size)
+            std::copy(ptr, ptr + size, data());
+         else {
+            _alloc<false>(ptr, size, size);
+         }
+      }
+
+      void assign(const std::vector<T>& v) {
+         assign(v.data(), v.size());
+      }
+
+      template<class... Args>
+      void emplace_back(Args&&... args) {
+         clear_and_construct(size() + 1, size(), [&](T* dest, std::size_t idx) {
+            std::construct_at(dest, std::forward<Args>(args)...);
+         });
+      }
+
+      const T* data() const {
+         return _data ? _data->data : nullptr;
+      }
+
+      T* data() {
+         return _data ? _data->data : nullptr;
+      }
+
+      std::size_t size() const {
+         return _data ? _data->size : 0;
+      }
+
+      bool empty() const {
+         return size() == 0;
+      }
+
+      const_iterator begin() const { return data(); }
+
+      const_iterator end() const {
+         return _data ? _data->data + _data->size : nullptr;
+      }
+
+      const_iterator cbegin() const { return begin(); }
+      const_iterator cend()   const { return end(); }
+
+      const T& operator[](std::size_t idx) const { assert(_data); return _data->data[idx]; }
+      T&       operator[](std::size_t idx)       { assert(_data); return _data->data[idx]; }
+
+      bool operator==(const shared_cow_vector& rhs) const {
+        return size() == rhs.size() && std::memcmp(data(), rhs.data(), size() * sizeof(T)) == 0;
```

Comment:
Shouldn't this use `std::equal`? What if the types have their own `operator==`?

Reply:
Good catch, done!